UI: Improved Button and IconButton

cdedreuille opened the pull request:
## What I did

As part of all our efforts to update our core UI components, we are improving both the Button and IconButton as part of this PR. The main concept here is to make it more of an atomic version that can be used anywhere. We worked on creating props that would make it easy to customise and extend the way you want without breaking the design. It also work with the existing icons library as well as our new icon library.
- No breaking changes for 8.0.
- One button to rule them all. We are removing all the need for special button instances.
- We are keeping the `children` prop as it is.
- new `asChild` prop to allow for a custom wrapper (for `<Link>` or `<a />`).
- `IconButton` is now just a wrapper around `Button` with special props.
- Match design of old props to the closest version we have in the new design.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

1. Open Storybook locally with `yarn storybook:ui`
2. Go to the IconButton story to test it out.

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

cdedreuille committed:
Add the new IconButton

cdedreuille committed:
Replace IconButton

cdedreuille committed:
Merge branch 'v8' into icon-button

cdedreuille committed:
Merge branch 'v8' into icon-button

cdedreuille committed:
Make IconButton works with text for backward compatibility

cdedreuille committed:
Remove the old IconButton

cdedreuille committed:
Update separator

cdedreuille committed:
Merge branch 'release-8-0' into icon-button

cdedreuille committed:
Fix icons in panel not showing correctly

cdedreuille committed:
Merge branch 'release-8-0' into icon-button

cdedreuille committed:
Update all props on IconButton

cdedreuille committed:
Merge branch 'release-8-0' into icon-button

cdedreuille committed:
Fix missing IconButton

cdedreuille committed:
Make icon prop optional

cdedreuille committed:
Update IconButton.tsx

cdedreuille committed:
Revert "Fix missing IconButton"

This reverts commit c9a9f06b3c742cbd05ea5c348a4dc3a3c7262f4b.

cdedreuille committed:
Revert

cdedreuille committed:
Fix some issues

cdedreuille committed:
Update IconButton.stories.tsx

cdedreuille committed:
Improved structure

cdedreuille committed:
Merge branch 'release-8-0' into icon-button

cdedreuille committed:
Improve Button

cdedreuille committed:
Replace IconButton with new Button component

cdedreuille committed:
Update eject.tsx

cdedreuille committed:
Cleanup

cdedreuille committed:
Update Button.tsx

cdedreuille committed:
Fix issues

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,29 +1,121 @@
 import { darken, lighten, rgba, transparentize } from 'polished';
+import { Slot } from '@radix-ui/react-slot';
+
+export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
```
`ButtonHTMLAttributes<HTMLButtonElement>` isn't strictly correct is it, as `asChild` could turn this into any element? @kasperpeulen 

I don't know if there are types here to inspire us.
https://github.com/radix-ui/primitives/blob/main/packages/react/primitive/src/Primitive.tsx

JReinhold replied:
Now that I've slept on it this might not be super important, given that props to the child should just be set on the child itself and not on the parent button. so while this is not technically true, it's probably okay.

kasperpeulen replied:
It is inspired by shadcn: https://ui.shadcn.com/docs/components/button

```tsx
export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  asChild?: boolean
}
 
const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : "button"
    return (
      <Comp
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...props}
      />
    )
  }
)
Button.displayName = "Button"
```

I think it is okay indeed.

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,29 +1,121 @@
-  ({ small, theme }) => ({
+}
+
+export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
```
Same question about `HTMLButtonElement` here

cdedreuille replied:
We talked about it with @kasperpeulen and it looked right but I defer to him on what is best here.

kasperpeulen replied:
Yes, also inspired by shadcn.

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,29 +1,121 @@
+      if (onClick) onClick(event);
+      if (onClickAnimation === 'none') return;
+      setIsAnimating(true);
+    };
```
```suggestion
    const handleClick = useCallback((event: SyntheticEvent) => {
      onClick.?(event);
      if (onClickAnimation === 'none') {
        return;
      }
      setIsAnimating(true);
    }, [onClick, onClickAnimation, setIsAnimating]);
```

(might not be 100% correct)

cdedreuille replied:
Not sure I understand the change here.

JReinhold replied:
1. wrap in `useCallback`
2. slightly easier syntax to read for `onClick`
3. no one-line if statements, but that's just a personal nit for me, feel free to ignore.

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,29 +1,121 @@
+      if (size === 'small') return '0 10px';
+      if (size === 'medium') return '0 12px';
+      return 0;
+    })()}`,
```
you don't need to wrap this in a string here.

```suggestion
    padding: (() => {
      if (padding === 'small' && size === 'small') return '0 7px';
      if (padding === 'small' && size === 'medium') return '0 9px';
      if (size === 'small') return '0 10px';
      if (size === 'medium') return '0 12px';
      return 0;
    })(),
```

cdedreuille replied:
Fixed it

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -33,240 +125,64 @@ const ButtonWrapper = styled.button<{
     whiteSpace: 'nowrap',
     userSelect: 'none',
-    opacity: 1,
+    opacity: disabled ? 0.5 : 1,
```
this way of setting something to be disabled is a bit unfortunate. if the button is on top of something that has multiple colors, it will actually be semi-transparent, where I think what we really want is the colors to be slighlty muted?

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -33,240 +125,64 @@ const ButtonWrapper = styled.button<{
-      path: {
-        fill: 'currentColor',
-      },
+    background: `${(() => {
```
same tip here about not needing the string wrapping.

cdedreuille replied:
Fixed

JReinhold replied:
it's not actually fixed here 😁

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -33,240 +125,64 @@ const ButtonWrapper = styled.button<{
-      marginTop: small ? 0 : -1,
-      marginBottom: small ? 0 : -1,
-
-      /* Necessary for js mouse events to not glitch out when hovering on svgs */
```
is this not needed anymore?

cdedreuille replied:
No. I made some tests and this is not needed

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -33,240 +125,64 @@ const ButtonWrapper = styled.button<{
+      minWidth: 14,
+      height: 14,
+      animation:
+        isAnimating && animation !== 'none' && `${theme.animation[animation]} 1000ms ease-out`,
```
```suggestion
        isAnimating && animation !== 'none' ? `${theme.animation[animation]} 1000ms ease-out` : '',
```

cdedreuille replied:
Fixed

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,29 +1,121 @@
+  onClickAnimation?: 'none' | 'rotate360' | 'glow' | 'jiggle';
 
-const ButtonWrapper = styled.button<{
+  /** @deprecated This API is not used anymore. It will be removed after 9.0. */
```
"This API is not used anymore." is not really true, since it is still respected - I think you can remove this sentence.

We also want to point the user towards the alternative that they can use instead.

Here's a format I would propose for all of these:

```ts
/**
 * @deprecated Use {@link asChild} instead
 */
```

It would also be nice to split `ButtonProps` into two types, `ButtonProps` and `DeprecatedButtonProps`, because it would make it more apparent to users what is being changed, but it's just my personal preference.

cdedreuille replied:
I improved the depreciated naming based on your feedbacks. I'm using `ButtonProps` in multiple places so I might keep it like this for now.

JReinhold commented on code/ui/components/src/components/IconButton/IconButton.stories.tsx:
```diff
@@ -0,0 +1,81 @@
+export default meta;
+type Story = StoryObj<typeof meta>;
+
+export const Base = {};
```
same comment here about separate stories.

JReinhold commented on code/ui/components/src/components/IconButton/IconButton.tsx:
```diff
@@ -0,0 +1,11 @@
+import type { ButtonProps } from '../Button/Button';
+import { Button } from '../Button/Button';
+
+export const IconButton = forwardRef<HTMLButtonElement, ButtonProps>(
```
shouldn't this be deprecated too? Or have I missed something in the confusion?

cdedreuille replied:
Well, I'm not sure. We could depreciate the full component indeed. To add an icon in the toolbar you need to do:
```tsx
<Button padding="small" variant="ghost">
  <Icons name="facehappy" />
</Button>
```
Maybe it's easier if we keep both so they can use it directly using:
```
<IconButton>
  <Icons name="facehappy" />
</IconButton>
```

JReinhold replied:
I see your point. I'm okay with keeping it.

cdedreuille replied:
Add a section on how to place button in toolbar

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,29 +1,121 @@
+
+export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
+  ({ asChild, onClickAnimation = 'none', onClick, ...props }, ref) => {
+    const Comp = asChild ? Slot : 'button';
```
shouldn't this be `'a'` if `props.isLink === true`?

cdedreuille replied:
Yes you're right. I just pushed a fix

cdedreuille committed:
Fix animation on svg

JReinhold reviewed (changes requested):
Great progress. Apart from my comments, here are some general input:

1. Using the deprecated props will not make the button look at all like it used to. Here are [the new deprecated stories](https://next--635781f3500dd2c49e189caf.chromatic.com/?path=/story/components-button--default), compare them with the [current stories for the button](https://635781f3500dd2c49e189caf-ttizcqwbtq.chromatic.com/?path=/docs/components-button-depreciated--docs) and there is a clear visual difference. I'm not sure if we should consider this a breaking change or not. The API is not broken, but how the button looks in the UI has definitely changed.

2. There are some thing missing here, mainly migration notes and runtime deprecation warnings. do you plan on adding this later, or as part of this PR? what other follow-up work are you planning to do, if any?

3. I can't get the animations to work in the stories, am I doing something wrong? I'm just opening the base button story and setting the `onClickAnimation` prop but that doesn't seem to do anything.

4. I think we should use this opportunity to make the stories here great, with proper descriptions for the components and the props, so the Button/Docs page is something actually useful. setting a trend for our future work on the UI.

cdedreuille committed:
Fix name

cdedreuille commented on code/ui/components/src/components/Button/Button.depreciated.stories.tsx:
Fixed

cdedreuille committed:
Fixed. I used Stack and Row

cdedreuille commented on code/ui/components/src/components/Button/Button.stories.tsx:
```diff
@@ -1,82 +1,129 @@
+  <div style={{ display: 'flex', flexDirection: 'column', gap: '1rem' }}>{children}</div>
+);
 
-export const Primary = { args: { primary: true, children: 'Primary' } };
-export const Secondary = { args: { secondary: true, children: 'Secondary' } };
-export const Tertiary = { args: { tertiary: true, children: 'Tertiary' } };
-export const Gray = { args: { gray: true, children: 'Gray' } };
+const Line = ({ children }: { children: ReactNode }) => (
```
Fixed. I used Stack and Row

cdedreuille commented on code/ui/components/src/components/Button/Button.stories.tsx:
```diff
@@ -1,82 +1,129 @@
-};
-export const DisabledGray = {
-  args: { disabled: true, gray: true, children: 'Disabled Gray' },
+export const Variants: Story = {
```
It is way easier to have them all together to test the different ones compared to have them all in different stories. I'm not convinced by the idea to separate all of them.

cdedreuille committed:
Removed string

cdedreuille committed:
Update Button.tsx

cdedreuille committed:
Update Button.tsx

cdedreuille committed:
use 'a' if we are using isLink

cdedreuille commented:
Thanks a lot for reviewing this PR @JReinhold. Here's some answers:

> Using the deprecated props will not make the button look at all like it used to. Here are [the new deprecated stories](https://next--635781f3500dd2c49e189caf.chromatic.com/?path=/story/components-button--default), compare them with the [current stories for the button](https://635781f3500dd2c49e189caf-ttizcqwbtq.chromatic.com/?path=/docs/components-button-depreciated--docs) and there is a clear visual difference. I'm not sure if we should consider this a breaking change or not. The API is not broken, but how the button looks in the UI has definitely changed.

This was a conscious decision. From the audit that I made so far I could find a single instance where any of the main button components were use in both internal and external addons. The idea is to make sure that automatically default to our new default design with the closest match.

> There are some thing missing here, mainly migration notes and runtime deprecation warnings. do you plan on adding this later, or as part of this PR? what other follow-up work are you planning to do, if any?

I'll work on that tomorrow on migration notes. I'm not sure what need to be done regarding runtime depreciation warnings though. Can you tell me more?

> I can't get the animations to work in the stories, am I doing something wrong? I'm just opening the base button story and setting the onClickAnimation prop but that doesn't seem to do anything.

This is now fixed. I added more stories to test animations.

> I think we should use this opportunity to make the stories here great, with proper descriptions for the components and the props, so the Button/Docs page is something actually useful. setting a trend for our future work on the UI.

Yes you're right. I can work on it tomorrow morning.

cdedreuille committed:
Fixes

cdedreuille committed:
Merge branch 'release-8-0' into icon-button

JReinhold commented:
I'm consistently seeing some weird pixel-rendering issues (in Chromium) around the corners, is it just me, or are we using an unusual combination of shadows/borders that does this?

![image](https://github.com/storybookjs/storybook/assets/5678122/6b3c2b58-81bf-44e7-af00-2b24af30708c)

JReinhold commented:
> This was a conscious decision. From the audit that I made so far I could find a single instance where any of the main button components were use in both internal and external addons. The idea is to make sure that automatically default to our new default design with the closest match.

That's fair, we should make sure to mention it in the migration notes.

> I'll work on that tomorrow on migration notes. I'm not sure what need to be done regarding runtime depreciation warnings though. Can you tell me more?

we should show a warning in the console when we detect that someone is using any of the deprecated props. See how we do it for `Description` here: https://github.com/storybookjs/storybook/blob/next/code/ui/blocks/src/blocks/Description.tsx#L143-L157

cdedreuille committed:
Add migration + deprecated warnings

cdedreuille committed:
Add isLink story

cdedreuille committed:
Improve deprecated warning

cdedreuille committed:
Improve Button docs

cdedreuille committed:
Improve docs

cdedreuille committed:
Merge branch 'release-8-0' into icon-button

JReinhold commented on MIGRATION.md:
```diff
@@ -320,6 +321,10 @@ In Storybook 7 it was possible to use `addons.setConfig({...});` to configure St
 
+#### New UI and props for Button and IconButton components
+
+We used to have a lot of different buttons in `@storybook/components` that were not used anywhere. We have removed them and added a new `Button` component that can be used in all places. The `IconButton` component has also been updated to use the new `Button` component. Going forward addon creators and Storybook maintainers should use the new `Button` component. `IconButton` is still available fo ...
```
We need to:

1. be specific about which exact components have been deprecated (currently it just says "a lot of different buttons") 
2. which exact props have been deprecated
3. which new props replace which old props
4. give an example of migrating from the old props to the new props

We don't need to specify stuff that we "might remove" in the future, it doesn't help the user and just creates uncertainty.

In reality this is mostly an accessible version of what's written in the source code and the types, but we can't expect users to read those at all. Consider that users will see the warning in the console with a link to this section, they'll need a way forward here to resolve the warning.

Here's a proposal:

> The X component from `@storybook/components` have been deprecated.
The following props to the `Button` component have been deprecated:
> - X use Y instead
> - X use Y instead
> - ...
> (same for `IconButton` if it has any deprecations)
> Here's an example of migrating from the old props to the new props:
> ```
> // deprecated usage
> <Button ... />
> 
> // new usage
> <Button ... />
> ```

cdedreuille replied:
I just pushed an update with some new notes and clear descriptions of what props are deprecated and what is the alternative.

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,272 +1,216 @@
+    let Comp: 'button' | 'a' | typeof Slot = 'button';
+    if (props.isLink) Comp = 'a';
+    if (asChild) Comp = Slot;
+    let lovalVariant = variant;
```
```suggestion
    let localVariant = variant;
```

cdedreuille replied:
Oops. Now Fixed

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,272 +1,216 @@
-          color,
+      deprecate(
+        `Use of deprecated props in the button ${
+          buttonContent.length > 0 ? `"${buttonContent[0]}"` : 'component'
```
would this work too?
```suggestion
          buttonContent.length > 0 ? `"${buttonContent.join(' ')}"` : 'component'
```

cdedreuille replied:
Yes. I committed the change

JReinhold commented on code/ui/components/src/components/Button/Button.tsx:
```diff
@@ -1,272 +1,216 @@
+  verticalAlign: 'top',
+  whiteSpace: 'nowrap',
+  userSelect: 'none',
+  opacity: disabled ? 0.5 : 1,
```
this way of setting something to be disabled is a bit unfortunate. if the button is on top of something that has multiple colors, it will actually be semi-transparent, where I think what we really want is the colors to be slighlty muted?

cdedreuille replied:
I agree with you on this one but I think this is something we might change after the theming v2.0 project. For now there is pretty low risk that the button will be below something else.

JReinhold reviewed (approved):
Looks good! I have a few minor comments but I'm sure you'll figure it out. Make sure to review all the UI Tests in Chromatic.

cdedreuille committed:
Improved migration notes

cdedreuille committed:
Fixed spelling mistake

cdedreuille committed:
Update Button.tsx

cdedreuille committed:
Cleaning

cdedreuille committed:
Merge branch 'release-8-0' into icon-button

cdedreuille committed:
Fix button shrinking
